### PR TITLE
[SPARK-8327] Fixes issue with Ganglia not available

### PR DESCRIPTION
**Pull request by smartkiwi**
Fixed libphp version to load the one available in AMI (Spark AMI: ami-5bb18832)

Fixed the issue when Ganglia is not available because:

Shutting down GANGLIA gmond: [FAILED]

Starting GANGLIA gmond: [ OK ]

Shutting down GANGLIA gmond: [FAILED]

Starting GANGLIA gmond: [ OK ]

Connection to <...> closed. <...> Stopping httpd:
[FAILED] Starting httpd: httpd: Syntax error on line 199 of /etc/httpd/conf/httpd.conf: Cannot load modules/libphp-5.5.so into server: /etc/httpd/modules/libphp-5.5.so: cannot open shared object file: No such file or directory [FAILED] [timing]

ganglia setup: 00h 00m 03s Connection to <...> closed.

Spark standalone cluster started at <...>:8080 Ganglia started at <...>:5080/ganglia

Done!

### CODE REVIEW

**Comment by smartkiwi** [comment]
It failed for me with spark 1.3.1
I've tested with with ec2/spark-ec2 script from master branch of the apache/spark - because spark 1.3.1 doesn't let me to specify custom spark-ec2 repo.

./spark-ec2 ... --spark-ec2-git-repo=https://github.com/smartkiwi/spark-ec2 --spark-ec2-git-branch=branch-1.3 -s 2 --instance-type=r3.large launch "my test cluster2"

**Comment by smartkiwi** [comment]
See related SO question:
http://stackoverflow.com/questions/30768454/cant-access-ganglia-on-ec2-spark-cluster

**Comment by shivaram** [comment]
Thanks @smartkiwi for the PR. Any idea what changed for this to be required ? AFAIK we haven't changed our AMI in a while

**Comment by smartkiwi** [comment]
@shivaram I see that ami ids were not updated for a year:
https://github.com/mesos/spark-ec2/tree/branch-1.4/ami-list

So it is either AMI itself were somehow changed (unlikely) or spark-ec2 started to use another image because of changed mapping (like instance type to AMI).

I was starting r3.large instance in default zone - and spark-ec2 have chosen ami-5bb18832 - which is listed in pvm file: https://github.com/mesos/spark-ec2/blob/c525694daf50986d7d072b612d8460a4d0d3d4d9/ami-list/us-east-1/pvm

But mapping says that r3.lage is mapped to hvm: https://github.com/apache/spark/blob/master/ec2/spark_ec2.py#L401

I've debugged spark-ec2.py -
with --instance-type=r3.large
It selected ami-35b1885c (hvm one) - which also had this issue - httpd failed to load libphp.so

And without instance-type set - it selected ami-5bb18832 (using default instance type m1.large)

I see that both amis has this issue.

**Comment by smartkiwi** [comment]
Is httpd preinstalled in ami image? Or installed after instance created?

**Comment by shivaram** [comment]
Hmm I think we install it from yum which might be the reason this changed [1]. It would be good to change that line to ask yum for a specific version of httpd.

BTW can you open this PR on branch-1.4 ? We typically merge things into the latest branch and then backport things if we don't run into any bugs etc.

[1] https://github.com/mesos/spark-ec2/blob/branch-1.4/ganglia/init.sh

**Comment by smartkiwi** [comment]
Ok. I've pinned httpd and php versions. Also to avoid future issues I propose to pin ganglia version too.
If there will be a new version of ganglia available - the version can be updated accordingly.

I've already tested this with branch-1.3

I'll submit pull request for branch-1.4 once I'll finish testing it

**Comment by smartkiwi** [comment]
Ok. I've finished testing with branch-1.4 - the branch uses AMI which comes with ganglia preinstalled. But the version of httpd and php are older (2.2 and 5.3) and have different issues. So my commit cannot be merged there - templates file has to be updated. Or AMI has to be rebuilt with newer version of Ganglia.

So I'm not going to open PR for branch-1.4 with my changes - they won't work with branch-1.4.

**Comment by shivaram** [comment]
Thanks @smartkiwi for investigating this.  BTW does ganglia work with the 1.4 AMI or does the libphp problem exist there as well ?

**Comment by smartkiwi** [comment]
Actually 1.4 AMI fail in other way :)

Starting httpd: httpd: Syntax error on line 154 of /etc/httpd/conf/httpd.conf: Cannot load /etc/httpd/modules/mod_authz_core.so into server: /etc/httpd/modules/mod_authz_core.so: cannot open shared object file: No such file or directory

It confused me a lot today when I tried my branch with spark-ec2 from spark 1.4.0. So I've debugged it - looks like ami already has ganglia preinstalled - so there is no need to install it with yum.
But the ganglia version installs older version of httpd and php. But httpd.conf should be updated accordingly.

I'm playing with it and will submit another PR

**Comment by smartkiwi** [comment]
I'm giving up on branch-1.4 it doesn't work properly - httpd doesn't start.
But also spark fails to install - /spark is empty

I've use the following command:
python ./spark_ec2.py -k <..> -i <...> -s 2 -z us-east-1e --spark-ec2-git-repo=https://github.com/smartkiwi/spark-ec2 --spark-ec2-git-branch=branch-1.4 launch "spark1.4_1"

with spark_ec2.py from the latest release.

Actually branch-1.4 had never worked for me - it also failed with spark_ec2.py from spark 1.3.1 release.

When I just us default spark-ec2 repository and default branch (branch-1.3) spark works, but ganglia doesn't - httpd fails to start with exception:
Starting httpd: httpd: Syntax error on line 154 of /etc/httpd/conf/httpd.conf: Cannot load /etc/httpd/modules/mod_authz_core.so into server: /etc/httpd/modules/mod_authz_core.so: cannot open shared object file: No such file or directory

**Comment by shivaram** [comment]
There were a couple of PRs that went in yesterday to make 1.4 work correctly https://github.com/mesos/spark-ec2/pull/119 and https://github.com/mesos/spark-ec2/pull/120 -- That might make the spark install problems go away

**Comment by smartkiwi** [comment]
@shivaram, aha that's explains it

I've just tried to launch cluster with my branch-1.3

python ./spark_ec2.py -k <...> -i <...> -s 2 -z us-east-1e --spark-ec2-git-repo=https://github.com/smartkiwi/spark-ec2 --spark-ec2-git-branch=branch-1.3 launch "spark1.4"

It failed to install spark too there - /spark is empty.

I'll use latest etc/2spark-ec2 from master then

**Comment by smartkiwi** [comment]
Ok. I've tested spark's 1.4 spark-ec2 with branch-1.3 and branch-1.4 of the mesos/spark-ec2.
I've created bug in the https://issues.apache.org/jira/browse/SPARK-8338
Looks like spark is using new AMI that has ganglia rpm and decencies installed back in 2013.
But templates/httpd/httpd.conf were updated.

So if we want to keep using preinstalled ganglia - we need to downgrade httpd.conf back to Feb 2013.

Personally I would suggest to recreate AMIs with newer versions of ganglia and httpd/php or at least update ganglia by removing old version and installing the new ones.
Basically that's what latest version of the pull request is doing.

**Comment by smartkiwi** [comment]
See newer related pull requests for spark 1.4
https://github.com/mesos/spark-ec2/pull/121
https://github.com/mesos/spark-ec2/pull/122

**Comment by smartkiwi** [comment]
The following pull request contains working solution
https://github.com/mesos/spark-ec2/pull/133
